[수달] 자동차 경주 step2 제출합니다.

otter66 opened the pull request:
step1 리뷰에서 말씀해주신 의견들을 고민해보며 
테스트 방법, 테스트 명명 규칙, 객체 검증 등을 특히 많이 찾아볼 수 있는 기회가 되었습니다! 
감사합니다!
아직도 여전히 부족하지만 말씀해주시는 부분들에 대해 직접 찾아보고, 스스로 개선해 나갈 수 있도록 노력하겠습니다!

tmdgh1592 committed:
docs: 기능 목록 초기 작성

tmdgh1592 committed:
feat: 최종 우승자 산출 기능 구현

tmdgh1592 committed:
feat: 자동차 이름들 입력 기능 구현

tmdgh1592 committed:
feat: 시도 횟수 입력 기능 구현

tmdgh1592 committed:
refactor: 조건 숫자, 에러문구 상수화

tmdgh1592 committed:
feat: 자동차 랜덤 전진 기능 구현

tmdgh1592 committed:
test: 테스트 커밋

tmdgh1592 committed:
feat: 실행 결과 출력 기능 구현

tmdgh1592 committed:
feat: 최종 우승자 산출 기능 구현

tmdgh1592 committed:
docs: 기능 목록 추가

tmdgh1592 committed:
feat: 우승자 출력 기능 구현

tmdgh1592 committed:
feat: 기능 통합 구현

tmdgh1592 committed:
test: Car class 테스트 초기 작성

tmdgh1592 committed:
test: Validator class 테스트 초기 작성

tmdgh1592 committed:
test: Validator class 시도 횟수 입력 테스트 초기 작성

tmdgh1592 committed:
test: RacingService class 객체 생성 테스트 작성

tmdgh1592 committed:
feat: 고정 시드 랜덤 유틸 구현

tmdgh1592 committed:
refactor: 랜덤 시작, 종료 범위, 이동 확률 상수화

tmdgh1592 committed:
test: 최종 우승자 산출 테스트 코드 구현

tmdgh1592 committed:
test: 최종 우승자 산출 테스트 코드 구현

tmdgh1592 committed:
test: 자동차 생성 예외 테스트 코드 구현

tmdgh1592 committed:
test: 최종 우승자 산출 예외 테스트 코드 구현

tmdgh1592 committed:
test: 전체 기능 테스트 코드 구현

tmdgh1592 committed:
feat: 공백 지우는 클래스 구현

tmdgh1592 committed:
refactor: 메서드명, 길이 10자 이내로 제한하도록 리팩토링

tmdgh1592 committed:
test: validator 테스트 코드 매개변수 수정

Commit:
refactor: 불필요 주석 제거

Commit:
refactor: 테스트코드 이름 변경

Commit:
refactor: BlankRemover class의 removeBlank를 TextUtils의 removeTextsBlank로 대체

Commit:
refactor: RacingController의 기능을 ViewController를 생성해 분할

Commit:
refactor: Validator 역할 축소, 및 이름 변경 (RacingRuleValidator)

Commit:
refactor: 상수 범위 축소

Commit:
refactor: 테스트명 변경

Commit:
refactor: 자동차 랜점 전진 기능 분할

Commit:
refactor: 자동차 랜점 전진 기능 분할

Commit:
fix: 요구사항과 일치하지 않는 수치 수정

Commit:
refactor: 테스트 명명 규칙 변경

Commit:
test: RacingService의 isMove메서드에 대한 테스트 구현

Commit:
refactor: Input 관련 내용을 ViewController로부터 InputController로 분리"

Commit:
refactor: output 관련 내용을 ViewController로부터 OutputController로 분리"

Commit:
test: Racing Service의 isMove 반환값 테스트 추가

Commit:
refactor: 메서드의 반환값 명시

Commit:
refactor: Car Class의 불변값인 name을 외부 참조 허용하도록 변경

Commit:
refactor: Random Class의 pickInRange Method의 코드를 간략화

Commit:
refactor: 테스트명 상세화

otter66 committed:
Merge branch 'otter66' into step2

Commit:
fix: git 충돌 해결

BeokBeok commented on src/main/kotlin/racingcar/controller/InputController.kt:
```diff
@@ -0,0 +1,45 @@
+import racingcar.view.InputView
+import racingcar.view.OutputView
+
+class InputController(
```
InputController와 OutputController는 Controller의 성격에 맞지 않아보여요.
이렇게되면, Controller가 다른 Controller들을 제어하는 상황이 생기게 됩니다.
Controller의 역할에 대해서 다시 한번 학습해보시고 코드를 개선해보면 어떨까요?

otter66 replied:
이전에는 값에 대한 검증을 View에서 하는 것에 대해 고민이 있어 Controller에서 검증을 하도록 하였습니다.
현재는 원시값에 대해 도메인을 만들어 값 검증을 도메인에서 하도록 변경하였고, 
InputController와 OutputController를 삭제해 View만 남겨두었습니다!! 리뷰 감사합니다!

BeokBeok commented on src/main/kotlin/racingcar/controller/InputController.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    companion object {
+        const val CAR_NAMES_REQUEST_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분)."
```
```suggestion
        private const val CAR_NAMES_REQUEST_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분)."
```
클래스 내부에서만 사용한다면, private 접근자를 붙여보면 어떨까요? (이하 관련 내용 동일)

otter66 replied:
감사합니다! private 접근자를 사용하였습니다!

BeokBeok commented on src/test/kotlin/racingcar/service/RacingServiceTest.kt:
```diff
@@ -21,36 +20,47 @@ internal class RacingServiceTest {
 
+    @ParameterizedTest
+    @ValueSource(ints = [9, 6, 4])
+    fun `발행된 전진 확률이 4~9일 때 isMove 메서드가 true를 반환합니다`(moveProbability: Int) {
```
```suggestion
    fun `발행된 전진 확률이 4이상일 때 isMove 메서드가 true를 반환합니다`(moveProbability: Int) {
```
"이상"과 "미만" 과 같은 단어를 활용해보면, 함수명만 보고도 로직을 이해하기가 편해져요 (이하 관련 내용 동일)

otter66 replied:
감사합니다! 이상, 이하, 미만, 초과 등의 단어를 사용하도록 수정하였습니다!

BeokBeok commented on src/test/kotlin/racingcar/utils/RacingRuleValidatorTest.kt:
```diff
@@ -40,15 +40,15 @@ internal class RacingRuleValidatorTest {
     @ParameterizedTest
     @ValueSource(strings = ["1", Int.MAX_VALUE.toString(), "5", "6"])
-    fun `시도 횟수 정상 테스트`(input: String) {
+    fun `시도 횟수가 1 ~ 2,147,483,647인 숫자일 때 예외가 발생하지 않습니다`(input: String) {
```
테스트 함수명에 제시된 숫자만 테스트 코드를 작성하는 것을 권장합니다. (이하 관련 내용 동일)

otter66 replied:
감사합니다! 필요 이상의 테스트케이스 제거와 함께 한가지 이상의 일을 하는 테스트 코드를 분리하였습니다!

BeokBeok commented on src/test/kotlin/racingcar/model/CarTest.kt:
```diff
@@ -13,14 +14,16 @@ class CarTest {
-
+    @ParameterizedTest
+    @ValueSource(ints = [0, 1, 2, 3])
+    fun `자동차가 전진한 횟수 만큼 "-"가 표현됩니다`(moveCount: Int) {
```
자동차 전진 횟수에 대한 테스트 코드를 잘 작성해주셨네요.
지금처럼 작성해주셔도 좋지만, 자동차 전진 여부와 관련된 변수는 position이니 position으로 비교해보면 어떨까요?

otter66 replied:
말씀해주신 것처럼 해당 테스트명은 전진 여부와는 별개의 테스트인 것 같습니다. 
 전진 여부 테스트와 "-"표현 테스트 분리하였습니다!

BeokBeok commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -32,6 +32,13 @@ class Car(
         return result
     }
 
+    fun move(isMove: Boolean) {
+        if (isMove) setPosition(position + 1)
+    }
```
아래 함수를 테스트하기 위해 만들어 놓은 함수로 보이네요.
다른 개발자의 입장에서 move함수를 호출할 때, 어떤 move함수를 호출해야할지 매번 선택의 순간이 올 것 같아요.
```kotlin
fun move() = ++position
```
가능하면 위 함수(move 함수)를 개선해보면 어떨까요?

otter66 replied:
말씀해주신 것처럼 move 함수를 하나로 만들고, 기능을 상세하게 설명할 수 있는 함수명으로 변경하였습니다!

BeokBeok reviewed (changes requested):
이전 단계 피드백과 자동차 경주 리팩터링 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

Commit:
refactor: 내부 사용 상수들의 접근자를 private으로 수정

Commit:
refactor: 테스트 코드명 가독성 향상
- 이상, 이하, 미만, 초과 등의 단어 사용
- 테스트케이스 분리
- 테스트 내용과 일치하도록 수정

Commit:
refactor: 자동차 이름 원시값 포장

Commit:
refactor: Car class의 move method 두 개를 takeOneStepForward로 통일

Commit:
refactor: Car class 패키지 변경

Commit:
"refactor: 자동차 이름 원시값 포장 해제"

Commit:
refactor: 불필요 파일 제거

Commit:
refactor: Cars 일급 컬랙션을 구현해 controller의 로직 축소

Commit:
refactor: 불필요 파일 제거

Commit:
refactor: InputController class가 controller의 성격에 맞지 않는 문제 해결

- RacingController class가 InputController class를 제어하지 않도록 수정

Commit:
refactor: OutputController class가 controller의 성격에 맞지 않는 문제 해결

- RacingController class가 OutputController class를 제어하지 않도록 OutputController의 역할을 OutputView가 할 수 있도록 수정

Commit:
refactor: 값 검증 위치 변경

TryCount class를 만들어 시도 횟수 값에 대한 검증을 InputView가 아닌 TryCount에서 할 수 있도록 변경

Commit:
refactor: 값 검증 위치 변경

RacingRuleValidator가 필요 이상의 의존성을 가진다고 판단해 Car에서 역할을 대신하도록 변경

Commit:
test: Car 객체에 대한 테스트케이스 추가

- "-"개수 테스트 분리
- 전진 기능에 대한 테스트 추가

BeokBeok commented on src/main/kotlin/racingcar/domain/TryCount.kt:
```diff
@@ -0,0 +1,18 @@
+import racingcar.utils.MAX_ROUND_COUNT
+import racingcar.utils.MIN_ROUND_COUNT
+
+class TryCount(
```
클래스명은 명사구로 작성하는 것을 권장합니다.

otter66 replied:
다시 보니 TryNumber와 같은 이름이 더 맞는 것 같습니다!
작성하고 일주일만 지나도 부족한점이 너무 많이 보이는 코드였지만.. 리뷰해 주셔서 감사드립니다!

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
좋은 개발자가 되기 위한 소중한 경험이 되었길 바랍니다. 💪
